Sema: Remove SourceFile::SynthesizedDecls

slavapestov opened the pull request.

slavapestov committed:
Sema: Remove SourceFile::SynthesizedDecls

slavapestov commented on test/Sema/enum_conformance_synthesis.swift:
```diff
@@ -61,7 +61,7 @@ func customHashable() {
   case A, B
 
-  var hashValue: String { return "" } // expected-error {{invalid redeclaration of synthesized implementation for protocol requirement 'hashValue'}}
+  var hashValue: String { return "" }
```
I wonder if this is OK. We'll still complain about ambiguity at the use site for hashValue, and since it has a different type the mangling for accessors is fine. (And if the user had defined it with the right type, we wouldn't try to synthesize it or complain at all)

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

DougGregor reviewed (approved):
Nice. I think the change in behavior is okay here.

CodaFi commented:
I think the change to auto diff is caused by their implicit expectation that the conformance is forced by the time they get to SILGen. I bet this is a maninfestation of that bug where they were using getMembers somewhere instead of getABIMembers.

slavapestov committed:
AutoDiff: Clean up getOrSynthesizeTangentVectorStruct() a bit

slavapestov committed:
AutoDiff: Use getLocalProtocols() instead of getInheritedProtocols()

slavapestov committed:
SILGen: Use getABIMembers() instead of getMembers() in a couple of spots

slavapestov committed:
SILGen: Assert that conformances are complete instead of skipping them

slavapestov committed:
AutoDiff: Synthesize the memberwise initializer earlier

We need to do this before we synthesize the body of the
AdditiveArithmetic witnesses, so that it becomes part of
the ABI members of the nominal.

slavapestov committed:
AutoDiff: Don't check hasInterfaceType() in ResolveEffectiveMemberwiseInitRequest::evaluate()

This check made this request depend on evaluation order, sometimes
synthesizing a second memberwise initializer if the previous one's
interface type had not been computed yet.

slavapestov committed:
AutoDiff: Check conformances on synthesized TangentVector struct

Once SF.SynthesizedDecls is gone, we can't rely on this happening for us.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test macOS
